src/main/java/com/flippingutilities/utilities/RecipeHandler.java

```diff
@@ -0,0 +1,146 @@
+        if (parentIdToRecipes.isPresent()) {
+            return this.parentIdToRecipes.get().containsKey(itemId);
+        }
+        return false;
+    }
+
+    public Map<Integer, Optional<FlippingItem>> getChildRecipeItems(int parentId, boolean isBuy, List<FlippingItem> items) {
+        Optional<Recipe> recipe = getApplicableRecipe(parentId, isBuy);
+        if (recipe.isEmpty()) {
+            return new HashMap<>();
+        }
+        Set<Integer> childrenIds = recipe.get().children.stream().map(recipeItem -> recipeItem.id).collect(Collectors.toSet());
+
+        Map<Integer, Optional<FlippingItem>> itemIdToChildren = new HashMap<>();
+        for (FlippingItem item : items) {
+            if (childrenIds.contains(item.getItemId())) {
+                itemIdToChildren.put(item.getItemId(), Optional.of(item));
+            }
+        }
+
+        childrenIds.forEach(id -> {
+            if (!itemIdToChildren.containsKey(id)) {
+                itemIdToChildren.put(id, Optional.empty());
+            }
+        });
+
+        return itemIdToChildren;
+    }
+
+    public Optional<Recipe> getApplicableRecipe(int parentId, boolean isBuy) {
+        if (!isRecipeParent(parentId)) {
+            return Optional.empty();
+        }
+        List<Recipe> combinationComponentsList = parentIdToRecipes.get().get(parentId);
+        RelationshipType relationshipType = isBuy? RelationshipType.BUY_PARENT: RelationshipType.SELL_PARENT;
+        return combinationComponentsList.stream().filter(cm ->
+                cm.relationshipType == RelationshipType.BIDIRECTIONAL || cm.relationshipType == relationshipType).findFirst();
+    }
+
+    /**
+     * Converts the flat CombinationMetadata list into a map from parent id to a list of CombinationMetadata.
+     * This is because a parent can have multiple CombinationMetadata associated with it, each defining a
+     * relationship with different child items.
+     */
+    private Optional<Map<Integer, List<Recipe>>> getParentIdToRecipes
+            (Optional<List<Recipe>> combinationMetadataOptional) {
+        if (combinationMetadataOptional.isEmpty()) {
+            return Optional.empty();
+        }
+        Map<Integer, List<Recipe>> parentToCombinationMetadata= new HashMap<>();
+        List<Recipe> combinationMetadata = combinationMetadataOptional.get();
+        combinationMetadata.forEach(cm -> {
+            if (parentToCombinationMetadata.containsKey(cm.parent.id)) {
+                parentToCombinationMetadata.get(cm.parent.id).add(cm);
+            }
+            else {
+                parentToCombinationMetadata.put(cm.parent.id, new ArrayList<>(List.of(cm)));
+            }
+        });
+
+        return Optional.of(parentToCombinationMetadata);
+    }
+
+    private Optional<List<Recipe>> loadRecipes() {
+        Request request = new Request.Builder()
+                .url("https://raw.githubusercontent.com/Flipping-Utilities/rl-plugin/master/data/recipes.json")
```

Comment:
Todo: Change master to a static version?
Either we can dynamically update the JSON without having to update RL but can't move the json, or the recipes are fixed.
Perhaps use https://github.com/Flipping-Utilities/osrs-datasets?

Reply:
Perfect, forgot we had that haha